Introduce a dedicated autosaves endpoint for handling autosave behavior

adamsilverstein opened the pull request:
## Description

Based on previous work from https://github.com/WordPress/gutenberg/pull/4156.

Leverages the final version of the Autosaves controller from @azaozz in Trac, https://core.trac.wordpress.org/ticket/43316#comment:76

See https://github.com/WordPress/gutenberg/issues/4112.

Does not implement the autosave alert when an autosave exists and the deletion of stale autosaves on load, these are handled in https://github.com/WordPress/gutenberg/pull/4218

## How Has This Been Tested?
Local testing, db monitoring. Verified behavior as expected, matches core.

* Create a new post, don't type anything
* Check the database, there will be a single auto-draft
* type a title and some content, pause for 10 seconds
* note the message 'Autosaving' ... 'Saved' ... 'Save draft'
* check the database, not that there is a single row for the post, with a draft status
* Click the 'save draft' button
* note the message 'Saving' ... 'Saved'  - button stays in saved state
* check the database, note a revision has been created because of the manual save action (not an autosave)
* publish the post
* check the database, note the post status is now published
* make some more changes and wait 10 seconds
* note the message 'Autosaving' ... then disappears (not save draft option) - @todo maybe should show saved briefly here?
* check the database, note an autosave revision has been created
* make some more changes and wait again for the autosave
* check the database and note the autosave is updated, and that there is only one autosave
* note that the 'update button' is available, click it
* note that 'update' is disabled (until you change the post)
* check the database, note that a new revision was created
* repeat the last three steps
* exit the post edit screen and log in as a different user, open the same post
* edit the post and wait for an autosave
* check the database, note that there are now two autosaves for the post, one for each user.

## Types of changes
* New WP_REST_Autosaves_Controller
* New REQUEST_POST_AUTOSAVE effect, autosavePost action
* Tracks `isPostAutosavable` separately from `isEditedPostSaveable`
* Leverage `withChangeDetection` to track autosave state (this needs more testing) 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

adamsilverstein committed:
autosaves with rest controller, first pass

adamsilverstein committed:
use newer title call

adamsilverstein committed:
autosave: compare raw attributes for change check

adamsilverstein committed:
Clean up autosave effects

adamsilverstein committed:
pull isAutosaving from props in PostSavedState

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
from .14

adamsilverstein committed:
effects, improve toggle autosave timing, enable autosaving for published posts

adamsilverstein committed:
Disable ‘save draft’ after draft is autosaved

adamsilverstein committed:
revert post saved state changes, keep save button active regardless of autosave status

adamsilverstein committed:
remove unrelated showAutosaveAlert

adamsilverstein committed:
fixes for eslint

adamsilverstein committed:
leverage withChangeDetection for autosaves

adamsilverstein committed:
Merge branch 'master' into fix/autosaves

# Conflicts:
#	editor/store/effects.js

adamsilverstein committed:
fixes for php slint

adamsilverstein commented:
I see some tests are failing, I'll get that fixed.

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

danielbachhuber reviewed (commented):
Excited to try this out! Left a few comments to start.

danielbachhuber commented:
> * note the message 'Autosaving' ... 'Saved' ... 'Save draft'
> * note the message 'Autosaving' ... then disappears (not save draft option) - @todo maybe should show saved briefly here?

See also #4644

Maybe we can steal behavior from elsewhere (Calypso, Google Docs)?

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into fix/autosaves

adamsilverstein committed:
rename WP_REST_Autosaves_Controller  -> GB_REST_Autosaves_Controller to avoid conflict with core class

adamsilverstein committed:
Add a docblock for savePost action

adamsilverstein commented on editor/store/actions.js:
```diff
@@ -353,9 +353,10 @@ export function editPost( edits ) {
 }
 
-export function savePost() {
+export function savePost( options ) {
```
added

adamsilverstein commented on lib/class-wp-rest-autosaves-controller.php:
```diff
@@ -0,0 +1,355 @@
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_Autosaves_Controller extends WP_REST_Revisions_Controller {
```
changed to GB_REST_Autosaves_Controller

adamsilverstein commented on lib/class-wp-rest-autosaves-controller.php:
```diff
@@ -0,0 +1,355 @@
+			return new WP_Error( 'rest_post_invalid_id', __( 'Invalid item ID.', 'gutenberg' ), array( 'status' => 404 ) );
+		}
+
+		return $this->parent_controller->update_item_permissions_check( $request );
```
the parent controller here is the post this autosave belongs to, and the permissions of the autosave decend directly from the parent post - this is similar to the approach we use on the revisions class.

adamsilverstein commented on lib/class-wp-rest-autosaves-controller.php:
```diff
@@ -0,0 +1,355 @@
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_items' ),
+					'permission_callback' => array( $this->revision_controller, 'get_items_permissions_check' ),
```
instantiating the parent sets it up correctly for the subsequent method calls.

adamsilverstein commented on lib/register.php:
```diff
@@ -431,6 +431,23 @@ function gutenberg_register_post_types() {
+	foreach ( get_post_types( array(), 'objects' ) as $post_type_object ) {
+		if ( ! empty( $post_type_object->rest_base ) ) {
+			if ( post_type_supports( $post_type_object->name, 'revisions' ) ) {
+				$autosaves = new WP_REST_Autosaves_Controller( $post_type_object->name );
```
sure assuming its straightforward to bring them over from core.

adamsilverstein committed:
check state.autosave exists before using

adamsilverstein committed:
Fix autosave tests

adamsilverstein committed:
Revert "leverage withChangeDetection for autosaves"

This reverts commit b189a3c942bb5d7c7038a20d0428c0de66867fe1.

# Conflicts:
#	editor/store/selectors.js

adamsilverstein committed:
rename rest controler file to match class passing phplint

adamsilverstein commented:
@aduth & @azaozz - can we get this in after a review so it can get some testing? if things work generally well, we can merge as is, or we can pull out the autosaves controller and commit that to core where it belongs. what do you think?

aduth reviewed (commented):
_(Sorry, I had review comments saved, but apparently had not submitted them before now. 😬 )_

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

mcsf commented on lib/class-gb-rest-autosaves-controller.php:
```diff
@@ -0,0 +1,355 @@
+ *
+ * @see WP_REST_Controller
+ */
+class GB_REST_Autosaves_Controller extends WP_REST_Revisions_Controller {
```
It's quite strange that a class _extends_ another but then _contains_ a instance of the latter. I get that it works; it's just conceptually a little harder to reason about.

adamsilverstein replied:
@danielbachhuber pointed this out earlier asking why we didn't just use the parent instead of instantiating the revisions controller. I'm open to changing this but found initialization was required for reusing the way I was trying to.

adamsilverstein commented on lib/class-gb-rest-autosaves-controller.php:
```diff
@@ -0,0 +1,355 @@
+	public function __construct( $parent_post_type ) {
+		$this->parent_post_type    = $parent_post_type;
+		$this->parent_controller   = new WP_REST_Posts_Controller( $parent_post_type );
+		$this->revision_controller = new WP_REST_Revisions_Controller( $parent_post_type );
```
nit-picky, but good point! - i'll fix

adamsilverstein replied:
fixed in https://github.com/WordPress/gutenberg/pull/6257/commits/93881fc1e5a633ba4a72b5a220fd9c3fad81ba8a

adamsilverstein committed:
Merge branch 'master' into fix/autosaves

danielbachhuber commented:
@adamsilverstein I'm open to polishing this as much as we can and iterating later. I'd really love to not see so many revisions being created.

adamsilverstein commented:
> @adamsilverstein I'm open to polishing this as much as we can and iterating later. I'd really love to not see so many revisions being created.

@danielbachhuber I'll try to address the feedback here so we can get this merged.

adamsilverstein committed:
set up isAutosaving

adamsilverstein commented on editor/components/post-publish-button/index.js:
```diff
@@ -25,6 +25,7 @@ export function PostPublishButton( {
 	visibility,
 	isPublishable,
 	isSaveable,
+	isAutosaving,
```
setting was lost in some recent changes, fixed in https://github.com/WordPress/gutenberg/pull/6257/commits/dad568397c81fc400d225b48e03419fa4d90f3b7

adamsilverstein committed:
rename controller WP_REST_Autosaves_Controller and only load if class doesn’t exist

adamsilverstein committed:
Cleanup doc blocks

adamsilverstein committed:
rename revision_controller -> revisions_controller

adamsilverstein commented on lib/class-gb-rest-autosaves-controller.php:
```diff
@@ -0,0 +1,355 @@
+	private $parent_controller;
+
+	/**
+	 * Parent controller.
```
fixed in https://github.com/WordPress/gutenberg/pull/6257/commits/501874833ce440da346386b066749727aefa926f

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into fix/autosaves

# Conflicts:
#	lib/load.php
#	lib/register.php

adamsilverstein committed:
clean up isAutosaving reducer, removing toggleAutosave action

adamsilverstein commented on editor/store/reducer.js:
```diff
@@ -584,6 +593,16 @@ export function isTyping( state = false, action ) {
 
+export function currentlyAutosaving( state = false, action ) {
+	switch ( action.type ) {
+		case 'DOING_AUTOSAVE':
```
removed in https://github.com/WordPress/gutenberg/pull/6257/commits/ed7d5539e657020fd3df21e8f4eb6d878202763c

adamsilverstein commented:
> @adamsilverstein I'm open to polishing this as much as we can and iterating later. I'd really love to not see so many revisions being created.

@danielbachhuber or @azaozz or @aduth - can you give this a test locally and verify the database records created are what you would expect - ie. nearly/exactly the same as for the classic editor, for both draft and published posts? I believe I have addressed every issue raised (except the revisions controller instantiation which we can handle later)and I'd love to get this merged.

adamsilverstein committed:
fixes for php-lint

adamsilverstein committed:
fix test: button stays in saved state when post not dirty

adamsilverstein committed:
restore state check

danielbachhuber commented:
I'm getting a 404 locally too:

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/36432/39388238-603c3706-4a33-11e8-9179-5aac7a47983d.png">

It doesn't look like `WP_REST_Autosaves_Controller` is ever instantiated?

danielbachhuber committed:
Properly instantiate the autosaves controller

danielbachhuber commented:
> It doesn't look like `WP_REST_Autosaves_Controller` is ever instantiated?

I've fixed this in a204ecec. However, I get this JavaScript error when I click "Save Draft" after the post has been auto-saved once:

> Error: Given action "REQUEST_POST_UPDATE", reducer "currentlyAutosaving" returned undefined. To ignore an action, you must explicitly return the previous state. If you want this reducer to hold no value, you can return null instead of undefined.

adamsilverstein commented:
> Error: Given action "REQUEST_POST_UPDATE", reducer "currentlyAutosaving" returned undefined. To ignore an action, you must explicitly return the previous state. If you want this reducer to hold no value, you can return null instead of undefined.

@danielbachhuber Argg, sorry. something must have been overwritten in a merge... Let me fix up the branch and get it working again.

adamsilverstein committed:
Merge branch 'fix/autosaves' of github.com:WordPress/gutenberg into fix/autosaves

adamsilverstein committed:
ensure isAutosave returns a boolean

adamsilverstein commented:
@danielbachhuber autosaving should work correctly after https://github.com/WordPress/gutenberg/pull/6257/commits/e5e9e2562ccec3cea2c8559e96261db6a0a35889 - turns out I broke it in https://github.com/WordPress/gutenberg/pull/6257/commits/ed7d5539e657020fd3df21e8f4eb6d878202763c addressing recent feedback.

danielbachhuber commented:
@adamsilverstein Can you make sure there are tests for it too?

adamsilverstein commented:
@danielbachhuber - yep let me pull in the tests from core, looks like some route related tests are failing as well.

adamsilverstein committed:
Add PHP Unit tests covering WP_REST_Autosaves_Controller functionality

adamsilverstein committed:
Avoid duplicate route resgistration

adamsilverstein committed:
Rename/fix up unit tests for rest controller

adamsilverstein commented:
@danielbachhuber added tests from core ticket and fixed up double route registration from https://github.com/WordPress/gutenberg/commit/a204ececefef5391d43e14781cd90a32b028cc3f

adamsilverstein committed:
fixes for phpcs

danielbachhuber commented:
Can we get a JS test for the error I ran into, too? It seems like the JS side of things could use better test coverage.

adamsilverstein commented:
I can work on some additional tests although autosaving does already have coverage (and is adjusted here), some of the new functionality is not covered. the error you encountered was a coding bug not sure how to test for that.

adamsilverstein committed:
Merge branch 'master' into fix/autosaves

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into fix/autosaves

# Conflicts:
#	editor/store/effects.js

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

aduth commented on editor/store/actions.js:
```diff
@@ -353,9 +353,17 @@ export function editPost( edits ) {
+/**
+ * Returns an action object to save the post.
+ *
+ * @param {Object} options Set autosave: true for an autosave.
```
Minor: JSDoc has a solution for documenting properties of an argument that we could leverage:

http://usejsdoc.org/tags-param.html#parameters-with-properties

adamsilverstein replied:
nice, changed in https://github.com/WordPress/gutenberg/pull/6257/commits/9142591bfba6ba3fdfb4ea81dc57e3dcb7024504

aduth commented on editor/store/effects.js:
```diff
@@ -313,16 +337,11 @@ export default {
-			//    the necessary REST API behavior does not yet exist
-			//  - May need to check for whether the status of the edited post
-			//    has changed from the saved copy (i.e. published -> pending)
+		if ( ! isEditedPostDirty( state ) ) {
```
With this, could we just eliminate the previous condition?

```js
if ( ! isEditedPostNew( state ) && ! isEditedPostDirty( state ) ) {
```

adamsilverstein replied:
yes, removed in https://github.com/WordPress/gutenberg/pull/6257/commits/6e7c57c2a9956e51d995e3e4737ffe286c400c2a

aduth commented on editor/store/reducer.js:
```diff
@@ -584,6 +593,18 @@ export function isTyping( state = false, action ) {
 	return state;
 }
 
+export function currentlyAutosaving( state = false, action ) {
```
Noting some naming inconsistencies: `saving` vs. `currentlyAutosaving` (`autosaving`?). Maybe fair because this one reflects a boolean value, whereas the other is a more complex object shape. But then for boolean values we also have `isTyping` vs. `currentlyAutosaving` (`isAutosaving`?).

adamsilverstein replied:
renamed to `isAutosaving` in https://github.com/WordPress/gutenberg/pull/6257/commits/5ee1caea9889e2134979e1b4ac05488cc82bb15b

aduth commented on editor/store/effects.js:
```diff
@@ -49,9 +49,7 @@ import {
 	getPostEdits,
-	isCurrentPostPublished,
 	isEditedPostDirty,
-	isEditedPostNew,
```
Why has this been removed? This is causing an error because the `AUTOSAVE` effect attempts to call it:

>Uncaught ReferenceError: isEditedPostNew is not defined

adamsilverstein replied:
that usage is now removed

aduth commented on editor/store/reducer.js:
```diff
@@ -228,6 +228,15 @@ export const editor = flow( [
 		ignoreTypes: [ 'RECEIVE_BLOCKS', 'RESET_POST' ],
 	} ),
 ] )( {
+	autosave( state = false, action ) {
```
Per https://github.com/WordPress/gutenberg/pull/6257#discussion_r183837142, I hesitate about including `autosave` here. Particularly, via `withChangeDetection`, `RESET_AUTOSAVE` would cause the post to be marked as dirty. I'm not sure if there are issues presently caused by this. (Should ideally be accounted for in E2E tests by #6209)

adamsilverstein replied:
Right! I moved this **back** out of the editor reducer section in https://github.com/WordPress/gutenberg/pull/6257/commits/724778061293c54cbb4db8416731bf3ea1314918

aduth commented on editor/store/reducer.js:
```diff
@@ -228,6 +228,15 @@ export const editor = flow( [
 		ignoreTypes: [ 'RECEIVE_BLOCKS', 'RESET_POST' ],
 	} ),
 ] )( {
+	autosave( state = false, action ) {
```
Why do we represent the default state as a boolean, rather than a more semantically meaningful empty value like `null` ?

adamsilverstein replied:
null is probably a better choice, chenged in https://github.com/WordPress/gutenberg/pull/6257/commits/193d32e36070d6a80e136d999500ee617b9d823e

aduth commented on editor/store/selectors.js:
```diff
@@ -301,6 +301,44 @@ export function isEditedPostEmpty( state ) {
+	// If the title, excerpt or content has changed, the post is autosavable.
+	if (
+		( autosave.title && title !== autosave.title.raw ) ||
+		( autosave.excerpt && excerpt !== autosave.excerpt.raw ) ||
```
It's not clear to me when `autosave` would have or not have these properties. For example, if `getEditedPostExcerpt` returned a non-empty value but `autosave.excerpt` were empty, this would not return `true`. Is that expected? Unit tests should clarify these expectations.

adamsilverstein replied:
I think autosave will always have the excerpt and title properties - they can't be empty, they can only be an empty string which will evaluate correctly. The reason for this is that the api always returns all fields, so excerpt is always returned even if "empty". autosave represents the last stored autosave so we can be confident an excerpt will always be set.  I'm going to remove the checks for now since they are confusing and I don't think they are needed

adamsilverstein replied:
fixed in https://github.com/WordPress/gutenberg/pull/6257/commits/35f41ee3c7363492e61ddbc7181955850376fd20

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into fix/autosaves

adamsilverstein committed:
document autosave as a parameter property for options

adamsilverstein committed:
remove redundant check for isEditedPostNew

adamsilverstein committed:
rename currentlyAutosaving -> isAutosaving

adamsilverstein committed:
move autosave reducer outside editor

adamsilverstein committed:
set the default autosave state to null

adamsilverstein committed:
remove unneded object check that confused logic

adamsilverstein committed:
fix for eslint

adamsilverstein committed:
ensure href set before using in gutenberg_add_target_schema_to_links

aduth commented on editor/store/actions.js:
```diff
@@ -353,9 +353,18 @@ export function editPost( edits ) {
+ * Returns an action object to save the post.
+ *
+ * @param {Object} options          Options for the save.
+ * @param {Object} options.autosave Perform an autosave if true.
```
`{Object}` should be `{boolean}`

adamsilverstein replied:
fixed in https://github.com/WordPress/gutenberg/pull/6257/commits/dbd3ab661dddef2398a9b2c77d03f58b633af3a2

aduth commented on editor/store/effects.js:
```diff
@@ -102,41 +100,66 @@ export default {
+				},
+				() => {
+					dispatch( {
+						type: 'RESET_AUTOSAVE',
```
Is the behavior here to revert back to the previous state when an autosave fails? This is already accomplished for the proper save via `redux-optimist`, and it seems like we could leverage the same here. We might not want to do it on a common `UPDATE_POST` (or do we?), but maybe a separate dummy action which includes the `optimist` meta? See below in the `else` case for `optimist` property handling.

adamsilverstein replied:
The reason we dispatch `RESET_AUTOSAVE` here is to set the 'isAutosaving' state to false (which in turn is reflected in the post publish button and post-saved-state display... see https://github.com/WordPress/gutenberg/blob/07cccd74f52814733142600ca0eec2abc460fa84/editor/store/reducer.js#L592

Is redux-optimist appropriate for this use case?

aduth replied:
As I recall, Redux optimist behaves in a way that an action dispatched can either be committed (confirmed) or reverted, in which case the original dispatched action is pretended to have never existed, though all other subsequent will proceed as expected. It could be relevant here, though seems like it could be just as well to reset explicitly.

adamsilverstein replied:
I think its fine to set this explicitly here. Optimist is really useful when you want to show the ui state completed (optimistically) then revert that if it somehow fails. In this case it seems fine to show the autosave is working until it completes or fails.

aduth commented on editor/store/reducer.js:
```diff
@@ -584,6 +584,18 @@ export function isTyping( state = false, action ) {
 	return state;
 }
 
+export function isAutosaving( state = false, action ) {
```
Can we add JSDoc describing what state is being tracked here.

adamsilverstein replied:
added in https://github.com/WordPress/gutenberg/pull/6257/commits/4204ad2dba24e3acc0797b9ae11fb178d2a442cf

aduth commented on editor/store/reducer.js:
```diff
@@ -1042,8 +1054,19 @@ export const blockListSettings = ( state = {}, action ) => {
 	return state;
 };
 
+export const autosave = ( state = null, action ) => {
```
Can we add JSDoc describing what state is being tracked here.

adamsilverstein replied:
added in https://github.com/WordPress/gutenberg/pull/6257/commits/dfe6bc3f235ec86d7764a5ebfeb54445b878276d

aduth commented on editor/store/reducer.js:
```diff
@@ -1042,8 +1054,19 @@ export const blockListSettings = ( state = {}, action ) => {
 };
 
+export const autosave = ( state = null, action ) => {
+	const { post } = action;
```
Minor: Why do we destructure `post`, but not `type` ? Why not both or neither?

adamsilverstein replied:
not sure why that wound up like that :( Removed the destruring to better match other reducers in https://github.com/WordPress/gutenberg/pull/6257/commits/6cd8729523a790f48901ffae30142c516cb082f4

aduth commented on editor/store/selectors.js:
```diff
@@ -300,6 +300,44 @@ export function isEditedPostEmpty( state ) {
+		( excerpt !== autosave.excerpt.raw ) ||
+		( content !== autosave.content.raw )
+	) {
+		return true;
```
Minor: This condition could be eliminated and replaced with a single `return` value:

```js
return (
	( title !== autosave.title.raw ) ||
	( excerpt !== autosave.excerpt.raw ) ||
	( content !== autosave.content.raw )
);
```

adamsilverstein replied:
nice! changed in https://github.com/WordPress/gutenberg/pull/6257/commits/7e8b6a23f528af3d984948aabf2a180230d5f567

aduth commented on editor/store/selectors.js:
```diff
@@ -300,6 +300,44 @@ export function isEditedPostEmpty( state ) {
+	}
+
+	// If we don't already have an autosave, the post is autosavable.
+	if ( ! hasAutosave( state ) ) {
```
Is a post autosavable if it has no content? Or are we relying on the consumer (e.g. `AutosaveMonitor`) to also check `isSavable` ? Should we want to impose that? If a post can't be autosaved without also being savable, should we just include that in `isPostAutosavable` ?

```js
if ( ! isPostSavable( state ) ) {
	return false;
}
```

adamsilverstein replied:
> Is a post autosavable if it has no content? 

Yes I think so, if thats a change (the post isn't new). For example, if you create a post with only a title and publish it, then edit the title and clear it. When autosave fires it will save the empty state as an autosave. I guess we could avoid this if thats desired, I can check the classic editor to see what the behavior is there for empty posts.

> If a post can't be autosaved without also being savable, should we just include that in isPostAutosavable ?

I couldn't find the `isPostSavable` or `isSavable` selectors, i think you mean `isEditedPostSaveable`

A post can't be saved if nothing has changed since the last save.
An autosave can be created - as long as the post has changed since the last autosave.

aduth commented on editor/store/selectors.js:
```diff
@@ -300,6 +300,44 @@ export function isEditedPostEmpty( state ) {
+ * @param  {Object}  state Global application state
+ * @return {boolean}       Whether the post can be autosaved
+ */
+export function isPostAutosavable( state ) {
```
Is it "saveable" or "savable"? 

We should pick one and keep consistency. Currently this is inconsistent with `isPostSaveable`

adamsilverstein replied:
Its either: https://en.oxforddictionaries.com/definition/saveable

Since you already have the shorter use, I'll update my PR with this variation.

adamsilverstein replied:
fixed in https://github.com/WordPress/gutenberg/pull/6257/commits/bb26c8cc580658e06dc0dcfc76525b6bf2c185fc

aduth commented on lib/rest-api.php:
```diff
@@ -18,6 +18,23 @@
+			continue;
+		}
+		$controller = new $class( $post_type->name );
+		if ( ! is_subclass_of( $controller, 'WP_REST_Controller' ) ) {
```
What purpose is this condition serving? To avoid registering the autosave route if the post type REST controller class is invalid? Can we leave a remark as an inline code comment?

Also, do we need to instantiate the class, or just pass the class name to `is_subclass_of` ?

```php
<?php

if ( ! is_subclass_of( $class, 'WP_REST_Controller' ) ) {
	continue;
}
```

http://php.net/manual/en/function.is-subclass-of.php

If we want to go one step further, we could take advantage of the fact that this function returns `false` if the class doesn't exist, so we could even eliminate the prior `class_exists` condition.

```
$ php -a
php > var_export( is_subclass_of( 'Foo', 'Bar' ) );
false
```

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into fix/autosaves

# Conflicts:
#	editor/store/reducer.js
#	lib/load.php
#	lib/rest-api.php

adamsilverstein committed:
include the autosaves controller

adamsilverstein committed:
docblock object -> boolean

adamsilverstein committed:
document isAutosaving reducer

adamsilverstein committed:
document autosave reducer

adamsilverstein committed:
skip destructuring post

adamsilverstein committed:
eliminate unneeded conditional

adamsilverstein committed:
rename autosavable -> autosaveable (with an e)

adamsilverstein committed:
Simplify class check

adamsilverstein committed:
fixes for phpcs

adamsilverstein committed:
savePost on CMD-S

adamsilverstein committed:
correctly return autosave from selector

adamsilverstein committed:
correct state autosave usage

adamsilverstein committed:
fix docblock

adamsilverstein committed:
add test: should stop autosave timer when the autosave is up to date

adamsilverstein committed:
correct test naming: should stop autosave timer when the autosave is up to date

adamsilverstein committed:
Add an e2e test for post autosaving

adamsilverstein committed:
whitespace

adamsilverstein commented:
Added unit (https://github.com/WordPress/gutenberg/pull/6257/commits/92b8d798bc85eea6c1ae156ee74de191686312bb) and e2e (https://github.com/WordPress/gutenberg/pull/6257/commits/74935ffb9c4fecc6547ec3eb95ccdd96852cd013) testing for autosaves.

adamsilverstein commented:
@aduth all tests are passing, going to give the functionality another spin: any additional code feedback?

adamsilverstein committed:
Merge branch 'master' into fix/autosaves

# Conflicts:
#	editor/components/post-publish-button/index.js

aduth commented on editor/store/effects.js:
```diff
@@ -102,41 +100,66 @@ export default {
+					} );
+
+					dispatch( {
+						type: 'REQUEST_POST_UPDATE_SUCCESS',
```
Why do we dispatch `REQUEST_POST_UPDATE_SUCCESS` for autosaves? It seems the action is used for two purposes:

- Displaying a notice, which we've explicitly opted out of for autosaves
- Resetting `isSaving` state, which appears to only be considered for full saves, not autosaves

adamsilverstein replied:
> As I recall, Redux optimist behaves in a way that an action dispatched can either be committed (confirmed) or reverted,

Right, I think its mostly useful where you want to show the UI in the updated state "optimistically" (eg adding a todo, you show the todo added), then if things fail for some reason you can show an error and roll back the change. Not sure how useful that is for the autosave?

adamsilverstein replied:
> Why do we dispatch REQUEST_POST_UPDATE_SUCCESS for autosaves? It seems the action is used for two purposes:

good question, let me dig in a bit further to see why this is here

aduth replied:
>Right, I think its mostly useful where you want to show the UI in the updated state "optimistically" (eg adding a todo, you show the todo added), then if things fail for some reason you can show an error and roll back the change. Not sure how useful that is for the autosave?

I think of it more in the pure data sense: We assume some success state where a new autosave will become the referenced value, and if a failure is to occur, we revert back to the prior state value.

adamsilverstein replied:
@aduth this is used in the `saving` reducer to reset the saving state. This call resets the state, and that gets reflects in the `PostSavedState` component - without this the saving action never completes.

I see how the naming is confusing, maybe I can try resetting `saving` on `RESET_AUTOSAVE` instead.

adamsilverstein replied:
@aduth In https://github.com/WordPress/gutenberg/pull/6257/commits/253110292d9e56776c003d4396a2e5c1d0b49c8b I switched this to resetting saving on the `RESET_AUTOSAVE`, eliminating the need to dispatch REQUEST_POST_UPDATE_SUCCESS

aduth commented on editor/store/selectors.js:
```diff
@@ -300,6 +300,40 @@ export function isEditedPostEmpty( state ) {
+	);
+}
+
+export function hasAutosave( state ) {
```
Can we add a JSDoc.

aduth replied:
Unit tests would be good too.

adamsilverstein replied:
👍

adamsilverstein replied:
@aduth added tests in https://github.com/WordPress/gutenberg/pull/6257/commits/c31eb055d276ea4fe44bbbd1764eecaaa3fad920

aduth commented on editor/store/selectors.js:
```diff
@@ -300,6 +300,40 @@ export function isEditedPostEmpty( state ) {
+/**
+ * Returns true if the post can be autosaved, or false otherwise.
+ *
+ * @param  {Object}  state Global application state
```
JSDoc standards treat `@param` and `@return` as separate groupings with newline between, no alignment necessary between groups.

https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions

adamsilverstein replied:
right, correcting

aduth commented on test/e2e/specs/change-detection.test.js:
```diff
@@ -52,6 +57,19 @@ describe( 'Change detection', () => {
 
+	it( 'Should autosave post', async () => {
+		await page.type( '.editor-post-title__input', 'Hello World' );
+		const isSavingState = await page.waitForSelector( '.editor-post-saved-state.is-saving' );
```
Is there a way you can think of that we could avoid sitting and waiting for the autosave to occur? Adds a significant delay to the runtime of E2E tests otherwise.

Not necessarily a blocker.

adamsilverstein replied:
In core we currently respect a config variable, `AUTOSAVE_INTERVAL` which gets localized as `autosaveL10n.autosaveInterval`. We probably want to respect the same [configuration option](https://codex.wordpress.org/Editing_wp-config.php#Modify_AutoSave_Interval) and it we follow the same pattern of localizing the setting, we could overwrite that here in our test to make it much shorter.

adamsilverstein replied:
Added support for `AUTOSAVE_INTERVAL` in https://github.com/WordPress/gutenberg/pull/6257/commits/d795e1db8ff040dced0afc6de6baace3521246ec

aduth replied:
It worries me a bit with race conditions. I do like the addition of respecting `AUTOSAVE_INTERVAL` constant though.

Could we just trigger the autosave programmatically?

```js
await page.evaluate( () => {
	dispatch( 'core/editor' ).doAutosave();
} );
```

adamsilverstein committed:
respect AUTOSAVE_INTERVAL for autosave timing

adamsilverstein committed:
reset saving state on `RESET_AUTOSAVE`

adamsilverstein committed:
add mu plugin to lower the autosave interval for testing

adamsilverstein committed:
docs spacing

adamsilverstein committed:
Add JSDoc for hasAutosave

adamsilverstein committed:
eslint: missing trailing comma

adamsilverstein committed:
Add tests for the hasAutosave selector

adamsilverstein commented:
@aduth Thanks for the feedback -  I believe I have addressed all of your feedback, please let me know if I missed anything.

aduth commented on editor/components/autosave-monitor/index.js:
```diff
@@ -35,13 +38,21 @@ export class AutosaveMonitor extends Component {
 			isDirty: isEditedPostDirty(),
 			isSaveable: isEditedPostSaveable(),
+			isAutosaveable: isPostAutosaveable(),
+			editorSettings: getEditorSettings(),
```
Minor: We should only need to pass in the specific value being used: `autosaveInterval`, for a few reasons:

- Tend to prefer simpler prop value types, as complex types (`{}` and `[]`) are prone to incurring more reconciliation than intended / necessary by strict inequality, where `{} !== {}` but `10 === 10`. 
   - I checked, and this isn't a concern here with the current implementation of `getEditorSettings`
- In the strictest sense of a presentational component, `AutosaveMonitor` doesn't care about settings as a whole, it cares specifically about an interval value, and the props passed in via `withSelect` should reflect these requirements.

Aside: The desire for a singular value to be passed highlights that it may be wise for a new singular `getEditorSetting` selector to be created, or even replacing the `getEditorSettings` function to avoid it being a source of future maintainability headaches (i.e. footgun).

adamsilverstein replied:
Good points and agrees with my sense that something could be better here. changed in https://github.com/WordPress/gutenberg/pull/6257/commits/2012003b091d4a6201b6208f33568e8842d4c2d6

aduth commented on editor/store/effects.js:
```diff
@@ -145,8 +161,9 @@ export default {
-		if ( ! isPublished && ! willPublish ) {
-			// If saving a non published post, don't show any notice
+
+		if ( isAutosave || ( ! isPublished && ! willPublish ) ) {
```
When will `isAutosave` be truthy ? (Now that we've removed `REQUEST_POST_UPDATE_SUCCESS` from the `isAutosave` flow above)

adamsilverstein replied:
right, good catch. removed in https://github.com/WordPress/gutenberg/pull/6257/commits/f88fe54a26075e9dd46b1d9d090f8a739aa39e31

aduth commented on editor/store/test/selectors.js:
```diff
@@ -3397,5 +3398,17 @@ describe( 'selectors', () => {
 		} );
+
+		it( 'should return false if no autosave is available', () => {
+			const state = {};
```
Is it possible for `hasAutosave` to be called with a state shape of `{}` in standard usage?

With an initial state, `state` would at least be `{ autosave: null }`.

Raising because there's a chance this test is not well-covering the implementation if, for example, the implementation was:

```
hasAutosave = ( state ) => undefined !== state.autosave
```

Where this test case would pass, but the selector itself would produce the wrong result on the initial state.

aduth commented on lib/client-assets.php:
```diff
@@ -1012,6 +1012,7 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 		'titlePlaceholder'    => apply_filters( 'enter_title_here', __( 'Add title', 'gutenberg' ), $post ),
 		'bodyPlaceholder'     => apply_filters( 'write_your_story', __( 'Write your story', 'gutenberg' ), $post ),
 		'isRTL'               => is_rtl(),
+		'autosaveInterval'    => defined( 'AUTOSAVE_INTERVAL' ) ? AUTOSAVE_INTERVAL : 60,
```
This has the impact of changing autosaves from 10 second delay to 60 second delay. While the latter may align with the classic editor, I think it's at least worth highlighting as a significant change to saving behavior, one which may be worth discussing separately in the #core-editor meeting or a separate pull request.

adamsilverstein replied:
True this does change the timing, happy to discuss further if you are concerned about this change.

In terms of highlighting changes from establish behavior - I think its the other way around, running autosaves at 10 second intervals would be a significant change in established behavior. The current saves in Gutenberg are not autosaves - they are in fact draft saves that create revisions and do not work with published posts.

aduth replied:
Regardless, it does not appear within the scope of this pull request to change the interval.

An argument might be made if it was a regression of behavior to implement autosaves with the current interval, but if anything, it's already an improvement over the previous state of Gutenberg.

It was raised because I know in the past there's been significant consideration of the UX of autosave and its role in Gutenberg as putting less of a burden on the user for explicitly saving by having autosaving which occurs more frequently in the background (cc @jasmussen). I don't know whether 10s or 60s or some other interval is necessarily the best, but it's something that shouldn't be changed on a whim.

adamsilverstein replied:
Agreed. If there is a design decision around increasing intervals for autosave to every 10 seconds, lets not change that. However, this is a significant change from current core behavior that is worth highlighting/documenting. Perhaps @jasmussen can help explain or give more background about why this interval was chosen.

Don't more frequent autosaves create visual distraction in the interface by constantly showing the 'Autosaving' Saved state and disabling/enabling the publish action?  I am also concerned about the increase in server load cased by the increased save frequency in Gutenberg.

adamsilverstein replied:
For reference, the 10 second delay was introduced in this PR: https://github.com/WordPress/gutenberg/pull/1945, I couldn't find a decision of why the delay is 10 seconds. Here is the current core autosave.js file where the interval is set: https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/js/autosave.js#L741

jasmussen replied:
As a preface, I would at any one point in time defer to technical considerations, i.e. the both of you. Not only in the name of shipping, but in the name of other deeper aspects that I might not have a fully technical grasp on, such as limiting the amount of revisions or drafts to something that makes sense.

As such, I'll give a high level overview of how I'd love to see WordPress handling saving, long-term. The short answer is: _like how Google Docs does it_. The explicit "Save" action should ideally be completely automatic, and not even a button the user has to press. As soon as a new document is created, it is _created_ (i.e. saved). Any change you make, is _saved_, ideally a few seconds after typing. The over-arching goal is that through modern technology we can spare you having to worry about losing your work. ⌘S is almost a nervous tick for anyone who grew up using Photoshop, of fear that things would crash. But we are moving towards place where "save" isn't something you need to worry about.

I'm aware of the technical issues, both with offline support, metaboxes, revisions, and so on. The fact that we do have an explicit Save button is an affordance that acknowledges that we cannot yet do without one. But any change that moves us closer to a potential future where we can remove the save button in favor of an autosave _message_, just like Google Docs, is a good step, in my book. The less the user has to think about, the better.

danielbachhuber replied:
I've dropped it back to 10 seconds for now 7fd7c79f9

adamsilverstein committed:
remove unused isAutosave in REQUEST_POST_UPDATE_SUCCESS

adamsilverstein committed:
adjust initial state for autosave test

adamsilverstein commented:
> Could we just trigger the autosave programmatically?

@aduth great idea, any idea how to get a handle on `dispatch` from within the tests? I didn't see an existing example... 

when i tried:

`import { dispatch } from '@wordpress/data';`

I got an error message running the test:
`Cannot find module '@wordpress/data' from 'change-detection.test.js'`

aduth commented:
>@aduth great idea, any idea how to get a handle on dispatch from within the tests?

One of the more confusing aspects of Puppeteer are the distinct contexts of Node (test assertions) and the browser, with communication occurring via the [DevTools protocol](https://chromedevtools.github.io/devtools-protocol/). The logic in the `page.evaluate` function callback is not run within Node; if you tried to access a variable defined immediately outside the scope of the function, it would fail. Instead, it's run as though it were within the headless browser page, meaning that variables from the tests aren't available (except when [passed in as `...args`](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pageevaluatepagefunction-args)), but also that globals in the browser _are_ available: Namely, `dispatch` (or `window.dispatch`).

Which is to say, I'd expect it should work as-is. Maybe adding `window.` to avoid the ESLint error.

adamsilverstein commented:
> Which is to say, I'd expect it should work as-is. Maybe adding window. to avoid the ESLint error.

Hmnmm, I think its more than a linter error?

![](https://cl.ly/172g461B2b43/Image%202018-05-14%20at%202.19.03%20PM.png)

adamsilverstein committed:
only pass autosaveInterval

aduth committed:
Testing: Trigger programmatic autosave

aduth commented:
I mean as-is with the snippet from my [previous comment](https://github.com/WordPress/gutenberg/pull/6257#discussion_r188005588), though it should have been `window.wp.data.dispatch`.

See 89131f1

adamsilverstein commented:
> should have been window.wp.data.dispatch.

ok, that works! I'll try to get the test fixed now (it failed).

adamsilverstein committed:
Merge branch 'fix/autosaves' of github.com:WordPress/gutenberg into fix/autosaves

danielbachhuber commented:
@adamsilverstein Anything I can do to help wrap this up? I'd love, love, love to see this landed.

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into fix/autosaves

# Conflicts:
#	editor/components/post-publish-button/index.js

adamsilverstein commented:
When I slow down the tests, I see `Uncaught TypeError: window.wp.data.dispatch(...).doAutosave is not a function` in the console

adamsilverstein committed:
revert 2012003

adamsilverstein committed:
remove unused `className`

adamsilverstein committed:
use isAutosaving

danielbachhuber committed:
Drop autosave interval back to 10 seconds, because 60 is really slow

danielbachhuber reviewed (approved):
👍 Love it!

aduth commented on editor/store/effects.js:
```diff
@@ -309,20 +326,11 @@ export default {
 			return;
```
The line above this: We check `isEditedPostSaveable`. Shouldn't this be `isEditedPostAutosaveable` ? Do we allow autosave to occur while autosave is already in progress? What happens then to the value of state's `isAutosaving` when the first autosave returns (i.e. triggers `RESET_AUTOSAVE`) but the second hasn't?

aduth replied:
With the changes in 7b4bb19, should the `AUTOSAVE` effect be checking `isEditedPostAutosaveable` instead of `isEditedPostSaveable` (which now includes `isEditedPostSaveable` anyways)?

aduth replied:
Maybe worth testing the behavior of <kbd>Cmd+S</kbd>, which [fires itself as `autosave`](https://github.com/WordPress/gutenberg/blob/master/editor/components/editor-global-keyboard-shortcuts/index.js#L128), where we rely on the effect to stop itself from an unnecessary save.

**Edit:** Particularly, what happens when I press <kbd>Cmd+S</kbd> in quick succession (second press before the first autosave completes).

aduth commented on editor/components/autosave-monitor/index.js:
```diff
@@ -6,10 +6,13 @@ import { withSelect, withDispatch } from '@wordpress/data';
+			prevProps.isSaveable !== isSaveable ||
+			prevProps.isAutosaveable !== isAutosaveable
+		) {
+			this.toggleTimer( isDirty && isSaveable && isAutosaveable );
```
I seem to recall raising previously, but am unable to find the conversation and resulting rebuttal (if one exists): When would `isAutosaveable` be `true` and `isSaveable` be `false` ? The issue I'm getting to here is that it seems like `isSaveable` should be built in to the selector logic for `isAutosaveable` and thus made redundant as a separate prop here, simplifying to `isDirty && isAutosaveable`.

adamsilverstein replied:
In this case `isSaveable ` maps to `isEditedPostSaveable` which tests if the post contains a title, an excerpt, or non-empty content. if you edit the content of an existing post and remove everything, the post is no longer savable, this prevents calling the autosave in that case. its possibly redundant to have this here and in `effects.js`, should i remove it there?

aduth replied:
Should `isEditedPostSaveable` be tested as part of the `isAutosaveable` selector? In other words, is it _ever_ the case that an autosave is valid where `isEditedPostSaveable` would otherwise return `false` ? This would allow us to simplify both the effect and this component, where we could simply test `isAutosaveable` and be assured that the post is also "saveable" (with valid content).

adamsilverstein replied:
sure, thats a good idea. I can move that test into `isAutosavable` to clear up the logic here.

adamsilverstein replied:
Updated in https://github.com/WordPress/gutenberg/pull/6257/commits/7b4bb191500c6bc81f372c75d6937d6472d0e24c, adding the `isEditedPostSaveable` logic to `isPostAutosaveable` and removing the extra calls to `isEditedPostSaveable` here

adamsilverstein committed:
clarify use of isEditedPostSaveable in effects/AUTOSAVE

adamsilverstein committed:
Merge branch 'fix/autosaves' of github.com:WordPress/gutenberg into fix/autosaves

adamsilverstein committed:
Prevent autosaves if an autosave is in progress.

adamsilverstein committed:
Add isEditedPostSaveable logic to isPostAutosaveable

adamsilverstein commented:
@aduth can we merge this and iterate? have i addressed all your feedback?

adamsilverstein committed:
autosave monitor tests - use isAutosaveable

adamsilverstein committed:
fix tests for autosave monitor

aduth commented on editor/components/autosave-monitor/index.js:
```diff
@@ -6,10 +6,17 @@ import { withSelect, withDispatch } from '@wordpress/data';
+
+		// Prevent autosaves if an autosave is in progress.
+		if ( isAutosaving ) {
+			return;
```
`return`ing here is not preventing, since it's the behavior of `toggleTimer` to either schedule or cancel a scheduled autosave. So if a `toggleTimer( true )` had been called previously and on the next `componentDidUpdate` `isAutosave` is `true`, the scheduled autosave won't be stopped.

Should we want this to be considered instead as part of the condition below? i.e.

```js
if (
	prevProps.isDirty !== isDirty ||
	prevProps.isAutosaveable !== isAutosaveable ||
	prevProps.isAutosaving !== isAutosaving
) {
	this.toggleTimer( isDirty && isAutosaveable && ! isAutosaving );
}
```

adamsilverstein replied:
that would work, although as you point out this only affects the setting of the timer. to truly prevent a duplicate request, we need to check if an autosaving is happening right before we try to initiate a new one.

aduth commented on editor/store/reducer.js:
```diff
@@ -584,6 +584,26 @@ export function isTyping( state = false, action ) {
+ *
+ * @return {boolean} Updated state.
+ */
+export function isAutosaving( state = false, action ) {
```
```js
it( 'should return ??? when the post is published while autosave is in progress', () => {
	const original = isAutosaving( undefined, savePost( { autosave: true } ) );
	const state = isAutosaving( original, savePost() );

	expect( state ).toBe( /* ??? */ );
} );
```

Note: UI limitations imposed aren't enough. If we want it to be a case prevented, at the very least it should be considered as part of an `isEditedPostPublishable` selector or equivalent state layer consideration (e.g. considered as part of the `REQUEST_POST_UPDATE` effect).

It maybe speaks to a broader concern of: Is having separate state handling for "is autosaving" vs. "is saving" causing us more headaches to handle than if they were one in the same? Does it make sense for them to be considered as equivalent?

adamsilverstein replied:
possibly. autosaves and saves are very distinct and an autosave could be happening and then a save fired, these shouldn't interfere with each other.  ideally, you should be able to click the update button even while an autosave is firing?

if you are concerned about race conditions between the two, we could drop isAutosaving and prevent saves during autosaves and autosaves during saves.

aduth commented on lib/class-wp-rest-autosaves-controller.php:
```diff
@@ -0,0 +1,355 @@
+			define( 'DOING_AUTOSAVE', true );
+		}
+
+		$post = get_post( $request->get_param( 'id' ) );
```
>We should rename `id` to `post`. `id` is typically a read-only value generated by the database.

Was this addressed @danielbachhuber ?

danielbachhuber replied:
I think we need to land this and iterate. I don't see my comment as blocking for the PR.

aduth commented on test/e2e/specs/change-detection.test.js:
```diff
@@ -52,6 +57,20 @@ describe( 'Change detection', () => {
+		const isSavingState = await page.waitForSelector( '.editor-post-saved-state.is-saving' );
+		const autosaveState = await ( await isSavingState.getProperty( 'innerText' ) ).jsonValue();
+
+		expect( autosaveState ).toBe( __( 'Autosaving' ) );
```
Related to https://github.com/WordPress/gutenberg/pull/6257#issuecomment-388903016 , the `__` is misleading as it's not doing anything for us. The evaluated value in the context of this file is based on the instance of `@wordpress/i18n` running in the Node process, whereas the value in the browser is based on the copy running in the browser window. If, theoretically, we were running a non-English locale in our test site, this would fail if that non-English locale data were not also set in the Node process. In other words, this appears as though we're comparing the two localized strings, but it's not providing any more benefit than just:

```js
expect( autosaveState ).toBe( 'Autosaving' );
```

adamsilverstein replied:
i'll remove the translation

aduth reviewed (dismissed):
Was 89131f1 simply reverted instead of fixed? I'm not comfortable with adding 10 seconds delay to every Travis build (the time it takes for `page.waitForSelector( '.editor-post-saved-state.is-saving' )` to resolve).

danielbachhuber commented:
@aduth To help expedite the process of landing this, can you submit a PR against this PR with your suggested improvements, and then we can move forward?

adamsilverstein commented:
> Was 89131f1 simply reverted instead of fixed? I'm not comfortable with adding 10 seconds delay to every Travis build (the time it takes for page.waitForSelector( '.editor-post-saved-state.is-saving' ) to resolve).

This is reduced to three seconds with an mu-plugin in the container - https://github.com/WordPress/gutenberg/pull/6257/commits/bd29ca00569d0656c397775d2c8ccbff82535e86#diff-f3775486fe31a5db9322b447a2c53bbaR10

adamsilverstein commented:
> With the changes in 7b4bb19, should the AUTOSAVE effect be checking isEditedPostAutosaveable instead of isEditedPostSaveable (which now includes isEditedPostSaveable anyways)?

this can probably be changed, yes

danielbachhuber commented:
@adamsilverstein This is already a vast improvement over the current implementation that generates a ton of revisions. I think we should land what you have and iterate.

adamsilverstein commented:
> Maybe worth testing the behavior of Cmd+S, which fires itself as autosave, where we rely on the effect to stop itself from an unnecessary save.

> Edit: Particularly, what happens when I press Cmd+S in quick succession (second press before the first autosave completes).

Yes, thats a great way to test - are you suggesting we should add that to the e2e tests?

adamsilverstein commented:
> @adamsilverstein This is already a vast improvement over the current implementation that generates a ton of revisions. I think we should land what you have and iterate.

I agree, maybe we can address any outstanding concerns in follow up Issue/PRs.

danielbachhuber commented:
> I agree, we can address any outstanding concerns in follow up Issue/PRs.

I've filed #4112 against `3.0` to track this.

It doesn't look like I can dismiss @aduth's review, so he'll need to update for merge.

I really think we should land this today (Friday 5/25) though, so it has at least a few days to soak prior to 3.0.

aduth committed:
Autosave: Cancel scheduled autosave when autosaving starts

aduth committed:
Autosave: Style: Use property shorthand for autosaveInterval assignment

aduth committed:
Revert "revert 2012003"

This reverts commit bd29ca00569d0656c397775d2c8ccbff82535e86.

aduth committed:
State: Rename `isPostAutosaveable` to `isEditedPostAutosaveable`

Consistency with `isEditedPostSaveable`. If we want to simplify, we must simplify both.

aduth committed:
State: Implement `isEditedPostAutosaveable` as composing `isAutosavingPost`

In case implementation details of detecting whether an autosave is in progress ever change.

aduth committed:
State: Implement autosave action as action creator

aduth committed:
State: Improve isAutosavingPost selector JSDoc

Newline between param, return. Full sentence stops. Grammatical consistency of "Returns ..." with other selectors.

aduth committed:
State: Revert doAutosave action creator name to autosave

If "save" is a verb, why not is "autosave" ?

aduth committed:
Testing: Avoid potential race conditions on Cmd+S save fail

Not guaranteed that notice will not already have been displayed by time Cmd+S press resolves.

aduth committed:
State: Implement autosaving as superset of saving condition

Simultaneous saves and autosaves should not be allowed, since it opens to race conditions where an explicit save _after_ and autosave occurs could be interpreted by the server _before_ the autosave.

aduth committed:
State: Consider save-in-progress as unsaveable

Move UI restriction to state consideration, evaluated as part of savePost effect handler.

aduth committed:
Testing: Evaluate autosave completion by class presence

Also eliminate possible race condition on autosave effecting selector prior to resolved evaluation.

aduth committed:
Keyboard Shortcuts: Disallow non-dirty save requests

UI restriction previously accounted for by fact that `AUTOSAVE` handler had aborted when post not dirty. This will need to be a future iteration on `savePost` effect.

aduth committed:
State: Prevent save request if post is not saveable

e.g. New post without content. Previously handled in `AUTOSAVE` handler since keyboard shortcuts executed `autosave` action dispatch.

aduth committed:
State: Track autosave object as subset of received properties

Only those used in isEditedPostAutosaveable implementation

aduth committed:
State: Remove redundant condition from autosaveable

Since already checking isEditedPostSaveable, which returns false if save (including autosave) is in progress

aduth committed:
State: Simplify isEditedPostAutosaveable as Array#some of fields

Consolidate to getEditedPostAttribute as canonical source of fields values

aduth committed:
State: Deprecate getEditedPostExcerpt selector

Consolidate to getEditedPostAttribute. Unlike content, there is no special behavior for excerpt. Even if there were, getEditedPostAttribute should be the entry point, which like content, would only then defer its implementation to getEditedPostExcerpt.

aduth commented:
I've pushed several commits in response to my own prior feedback. Most include extended commit descriptions with more explanation.

At a high-level, the goals were:

- Consolidate "saving" and "autosaving" behaviors, both in effect handling and in representation by state / selectors (where autosave is reflected as a superset of the condition of saving).
- Restore (and resolve failings of) immediate autosave dispatch in end-to-end tests.
   - I did observe a test failure which resolved itself after restarting build. Related: #6956 
- Simplify representation of autosave object in state. It appears we only ever consider its `title`, `content`, and `excerpt`, so we don't need to track the entire post object.
- Add missing tests.
- Fixes.
   - Prior to these changes it was possible to Cmd+S to save a new post, which would fail. (End-to-end test case added)
   - Prior to these changes it was possible to Cmd+S to save a post which already had been saved, which would succeed but is not an intended supported behavior (and did not exist previously). (End-to-end test case added)
- Marked TODO where I observed issues but did not introduce changes (to avoid scope creep)
   - https://github.com/WordPress/gutenberg/pull/6257/files#diff-d5c01e166c630687dcb047cc4167d8dcR126
   - https://github.com/WordPress/gutenberg/pull/6257/files#diff-4e8871e471c420334a2b54cde7391e07R49
   - https://github.com/WordPress/gutenberg/pull/6257/files#diff-4e1dd4776f75eb7a66b28280a56d8e0aR269
   - https://github.com/WordPress/gutenberg/pull/6257/files#diff-4e1dd4776f75eb7a66b28280a56d8e0aR290
- Open questions:
   - What's the relationship between autosave and saving meta box values? This has not been changed from its previous logic, where metabox updates [trigger in response to the start of a save request](https://github.com/WordPress/gutenberg/blob/1e689af43b6e29336678eac29278de8afaf5808c/edit-post/store/effects.js#L55-L60) (including autosave).

I will dismiss my own review. If all appears well to you, feel free to merge.

adamsilverstein commented:
@aduth Thanks for all your work here, I will give this a test to verify the autosaves are still working as expected. I see there is a merge conflict now, I'll see if I can resolve that.

> What's the relationship between autosave and saving meta box values? 

in the current core implementation, metabox data _is not included_ in autosaves, however the heartbeat transport mechanism is hookable in a way that enables developers to add meta box data to autosaves on their own if they want. for regular (draft/update) saves, all of the metabox data is sent as part of the <form submission and developers can then store/respond on the PHP side. To match that here, we can probably exclude that triggering for autosaves.

danielbachhuber committed:
Merge branch 'master' into fix/autosaves

danielbachhuber commented:
> I see there is a merge conflict now, I'll see if I can resolve that.

Fixed in 0bec4eae1d488da6fba275eef04bbaae907ddcc1

danielbachhuber committed:
Remove tests originally removed in bdcbf714345045b5c96ef2d61ee986a1cf26fe0f

These ended up on the wrong side of the merge conflict 0bec4eae1d488da6fba275eef04bbaae907ddcc1

danielbachhuber committed:
Fix linting

danielbachhuber commented:
Green means go! @pento said he'd clean up any fallout 😝

adamsilverstein commented:
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 

Thanks @danielbachhuber - I'll also pick up any issues!

aduth commented:
@danielbachhuber Can you please direct me to the issue and/or pull request where the following promised iterations are addressed?

- https://github.com/WordPress/gutenberg/pull/6257#discussion_r190731872

danielbachhuber commented:
> Can you please direct me to the issue and/or pull request where the following promised iterations are addressed?

https://core.trac.wordpress.org/ticket/43316#comment:82

We'll make sure it's polished top to bottom before it lands in core.

mtias commented:
@adamsilverstein thanks for driving this one!
